xcube_clms/clms.py

```diff
@@ -18,16 +18,263 @@
 # LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
+import inspect
+import os
+from typing import Any, get_args, Container
 
 import xarray as xr
+from xcube.core.store import DataTypeLike
+from xcube.util.jsonschema import JsonObjectSchema, JsonStringSchema
 
+from .constants import (
+    SEARCH_ENDPOINT,
+    LOG,
+    CLMS_DATA_ID_KEY,
+    DOWNLOADABLE_FILES_KEY,
+    ITEMS_KEY,
+    CRS_KEY,
+    START_TIME_KEY,
+    END_TIME_KEY,
+    JSON_TYPE,
+    BATCH,
+    NEXT,
+    SCHEMA_KEY,
+    PROPERTIES_KEY,
+    ALLOWED_SCHEMA_PARAMS,
+    FILE_KEY,
+    DATA_ID_SEPARATOR,
+    ITEM_KEY,
+    PRODUCT_KEY,
+)
+from .preload import PreloadData
+from .utils import (
+    is_valid_data_type,
+    make_api_request,
+    get_response_of_type,
+    build_api_url,
+)
 
-class CLMS:
 
-    def __init__(self, url: str):
+class CLMS:
+    def __init__(self, url: str, credentials: dict, path: str | None = None):
         self._url = url
+        self._datasets_info: list[dict[str, Any]] = []
+        self._metadata: list[str] = []
+        self._data_ids = []
+        if path is None:
+            self.path = os.path.join(os.getcwd(), "preload_cache/")
+        else:
+            self.path = os.path.join(os.getcwd(), path)
+        self._preload_data = PreloadData(self._url, credentials, self.path)
 
-    def open_dataset(
-        self, data_id: str, **open_params
+        self.file_store = self._preload_data.file_store
+        self._fetch_all_datasets()
+
+    def open_data(
+        self,
+        data_id: str,
+        **open_params,
     ) -> xr.Dataset:
-        raise NotImplementedError
+        try:
+            _, file_id = data_id.split(DATA_ID_SEPARATOR)
+        except ValueError as e:
+            raise ValueError(
+                f"Expected a data_id in the format {{ product_id}}"
+                f"{DATA_ID_SEPARATOR}{{file_id}} but got {data_id}"
+            )
+        if self.file_store is None:
+            raise ValueError(
+                "File store does not exist yet. Please preload "
+                "data first using the preload_data() method."
+            )
+        self._preload_data.refresh_cache()
+        if data_id in self._preload_data.cache.keys():
+            folder = self._preload_data.cache[data_id]
+            print(folder)
+            data_id_file = os.listdir(folder)
+            print(data_id_file)
+            if len(data_id_file) != 1:
+                LOG.error(
+                    f"Expected 1 file in the folder {folder}, "
+                    f"got {len(data_id_file)}. Opening the first file."
+                )
+            return self.file_store.open_data(
+                os.path.join(data_id, data_id_file[0]), **open_params
+            )
+
+    def get_data_ids(
```

Comment:
They do. But not here. The user will use the store interface and not the CLMS interface. In the store interface, I return an iterator.

Reply:
In the store interface, a list of all data IDs will be created by calling `get_data_ids` from the `CLMS` class. This can take very long, if a lot of datasets are available within one data store. An iterator only makes sense, when the actual work is performed during iteration. 

I think you just need to exchange a small piece in `CLMS._create_data_ids`. Here you create a list within a for-loop. Instead return a iterator here as well.